src/elements/Container/Container.js

```diff
@@ -1,33 +1,68 @@
-import React, { Component, PropTypes } from 'react'
-import classNames from 'classnames'
-import META from '../../utils/Meta'
+import cx from 'classnames'
+import React, { PropTypes } from 'react'
+import META from '../../utils/Meta';
+import { getUnhandledProps, useKeyOnly } from '../../utils/propUtils'
 
 /**
- * A container that gives your content some side padding.
+ * A container limits content to a maximum width
  */
-export default class Container extends Component {
-  static propTypes = {
-    children: PropTypes.node,
-    className: PropTypes.string,
-  }
-
-  static _meta = {
-    library: META.library.semanticUI,
-    name: 'Container',
-    type: META.type.element,
-  }
-
-  render() {
-    const classes = classNames(
-      'sd-container',
-      'ui',
-      this.props.className,
-      'container',
-    )
-    return (
-      <div {...this.props} className={classes}>
-        {this.props.children}
-      </div>
-    )
-  }
+function Container(props) {
+  const {
+    text, left, center, right, justified, fluid,
+    children, className,
+  } = props
+
+  const classes = cx('sd-container ui',
+    useKeyOnly(text, 'text'),
+    useKeyOnly(left, 'left aligned'),
+    useKeyOnly(center, 'center aligned'),
+    useKeyOnly(right, 'right aligned'),
+    useKeyOnly(justified, 'justified'),
```

Comment:
Understood.  But with only one possible instance of the 'aligned' prop 
being possible, be it 'left', 'center' or 'right', we'd only require one 
useValueAndKey mapping rather than 3?

I keep finding the api in terms of what should be Common props as enums, 
and bool for others, a little erratic.  Some props feel as though they 
should be one way, but only work the other.

For example, Button actions.  It feels right that you'd declare a 
<Button action/> but this breaks down when stipulating that it should be 
left or right <Button left action/>, but would seem even worse when 
combining it with an Icon:

<Button left action right icon/>

It seems like some commonality is required across all component props, 
as the same would apply for <Container className="left align"/>

Might defined props make things clearly for the end user?

<Container leftAlign />
<Container justify />
<Button leftAction />
<Button rightIcon/>

Or indeed, imposing defined props types with oneOf enums:

<Button align="left|center|right|justify" />
> Levi Thomason <mailto:notifications@github.com>
> 24 June 2016 at 17:23
>
> In src/elements/Container/Container.js 
> <https://github.com/TechnologyAdvice/stardust/pull/277#discussion_r68423298>:
>
> > - {this.props.children}
> > - </div>
> > - )
> > - }
> > +function Container(props) {
> > + const {
> > + text, left, center, right, justified, fluid,
> > + children, className,
> > + } = props
> > +
> > + const classes = cx('sd-container ui',
> > + useKeyOnly(text, 'text'),
> > + useKeyOnly(left, 'left aligned'),
> > + useKeyOnly(center, 'center aligned'),
> > + useKeyOnly(right, 'right aligned'),
> > + useKeyOnly(justified, 'justified'),
>
> There are 4 types of possible class to props mappings for SUI classes. 
> The alignment class here follows this API pattern:
>
> aligned="left" // useValueAndKey(aligned, 'aligned')
> aligned="right" // useValueAndKey(aligned, 'aligned')
> aligned="center" // useValueAndKey(aligned, 'aligned')
>
> One reason we use this key/value relationship is because two 
> simultaneous alignment props are not valid:
>
> |<Container right left />
> |
>
> Where as making those values of the |aligned| prop means you can only 
> have one at a time:
>
> |<Container aligned="left" />
> |
>
> However, the |justified| option breaks down however since the 
> resulting class is not |justified aligned|. We've not yet run into 
> this matchup yet so I'm open to suggestions. The best immediate 
> thought I have is to still allow |aligned="justified"| for the API 
> consistency, then in cx buildup:
>
> cx(
> useValueAndKey(aligned, 'aligned'),
> useValueAndKey(aligned, 'aligned'),
> useValueAndKey(aligned, 'aligned'),
> aligned === 'justified' && 'justified'
> )
>
> Thoughts?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/TechnologyAdvice/stardust/pull/277/files/f420da9c4eb10853132c436ba6a06faa82ffeba2#r68423298>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AAWHs4uiOF121pijncyPFGMdsccwEtpJks5qPARsgaJpZM4I9hiP>.
>

-- 
Sent from Postbox 
<https://www.postbox-inc.com/?utm_source=email&utm_medium=siglink&utm_campaign=reach>